### DISCUSSION TITLE
improving pcmanfm

### DISCUSSION THREAD

**Discussion by hektr** [lxqt/lxqt#2025] [Ideas]
It would be great to steal some ideas from ranger file manager and use them on pcmanfm. For example, one of the most useful ideas is tagging files. Here are example, demonstrating, what is it and why is it useful.

Suppose you have 300-400 pics in folder, but you need only 30 best pics from this list. Now you need to do the next actions on each pic:
1. open or preview pic
2. decide, whether it is good one or not
3. copy/paste or skip(?delete)

Now, let's see, how you can do the same thing with ranger:
1.preview file
2. decide, whether it is good one or not
3. tag file

At the end you just mark all tagged files(with command :mark_tag TAG) and can do with this selection, whatever you want (delete, copy-paste, etc). You can also use various tags, if you are not sure about quality of pic. For example, use tag 5 for good pics, 4 for not so good pics, 3 for bad pics, etc. And then, if you want to delete files with tag 3, just select them, using `:mark_tag 3` and then delete them using `:delete`

It would be great to see similar functionality in pcmanfm :) What do you think???

**Comment by tsujan** [comment]
pcmanfm-qt already supports emblems of any kind. The emblems can be added by custom actions. You could visually mark files by adding them.

The documentation about custom actions is still missing but I have a plan to add it to https://github.com/lxqt/pcmanfm-qt/wiki. I'll use emblems as a practical example. But I should find the time first ;)

There can be many valid features an app may lack. But these questions are also valid:
 * Is adding them worth the effort? How much work is required?
 * Will the app be still lightweight with them (that's important to us)?
 * What's the probability of introducing new bugs alongside them and how serious they could be?
 * Can the functionality be already achieved in other ways?

And more...

Please note that I'm not saying the the feature you mentioned isn't good.

**Comment by stefonarch** [comment]
For tagging pictures I use digikam (but shotwell and maybe others have the same feature of "virtual ordering"), and those tags are not limited to a folder but can include any picture in any folder in the database. 
So for pictures I don't think this is a good idea, reading exif data and extract tag is not compatible to lightweight probably.
But for other files it could be interesting, I remember dolphin and KDE going for  the "semantic" desktop years ago but it doesn't look to me that it was real followed by users.

**Reply by tsujan** [reply]
"Tags" are nothing but metadata, which are fully supported by GVFS/GLib (and pcmanfm-qt is based on GLib). Emblems are just one example of metadata.

Theoretically, by using command-line, you could add any metadata to files and find them later. However, the more metadata is supported by *the file manager* (actually, by `libfm-qt`), the more CPU time will be taken for loading heavy folders.

**Reply by tsujan** [reply]
BTW, metadata have two extra properties:

 1. Adding/removing them does nothing to the real file. They're stored in a special place that we don't need to worry about.
 2. They aren't limited to certain mimetypes (images, videos, etc.).

However, they have a cost: the app must search for and find them for every file.

**Reply by hektr** [reply]
I don't think, it is 'real' metadata in ranger. Tag in ranger is just a single character. Ranger is a console file manager and commands are performing on selected files there. So, tag is just a method to somehow 'mark' files without selecting them.. It allows you to perform commands on other files(=view other pics) and fast select tagged files when they are needed you. Reading all metadata, especially from images will make file manager working very slow.

May be, it is possible to make for such 'tags' some hidden json file in any directory, if that directory contains files 'tagged' by user. I think, it does't make file manager working slow, because program will take a list "file:tag" from that json file and doesn't need to check every file for tags. To keep things simple it's possible to use only default tag  `*` and then see, whether people use this feature or not.

**Reply by tsujan** [reply]
> I don't think, it is 'real' metadata in ranger. Tag in ranger...

`ranger` may have a non-standard way of "tagging" specific to itself. Instead, we use methods that are shared by a good variety of file managers.

**Reply by hektr** [reply]
May be embems is good for the same purpose🤔🙂

**Reply by tsujan** [reply]
> May be emblems is good for the same purpose

They're already supported. We should either tell users how to use them or add a GUI for them.

I'll try to add a button for them to the file properties dialog. In that way, it could also be used in LXQt file dialog and won't clutter the GUI of pcmanfm-qt. There will be no extra cost either.

**Reply by tsujan** [reply]
Examples of emblems in pcmanfm-qt — they can be seen by all good-enough GTK file managers too (like Thunar and Nautilus):

![emblems](https://user-images.githubusercontent.com/981076/123901625-82e4b500-d980-11eb-8336-0fb3821f331c.png)

**Comment by stefonarch** [comment]
Instead of opening a new one I add this here. 
Pcmanfm-qt has filtering and this possibility to mark virtually files or directories as "hidden" by inserting the filename(s) in a file `.hidden` in the same directory.

Now I was wondering if `.hidden` supports wildcards, and no, actually it doesn't.
Example: raw file along  with their xmp files and jpg files in the same directory (yy/mm/):
```
*.PEF
*.xmp*
```
@tsujan Would this be a) easy  to implement  or would there be b) drawbacks with many files or else?

**Reply by tsujan** [reply]
Good idea. We should first see whether other file managers that support `.hidden` also support wildcards in it. If they don't, we don't want a discrepancy; if they do, we'll need to evaluate the cost (resource usage) for ourself.

**Reply by tsujan** [reply]
Minutes ago, I installed Nautilus to test. It supported `.hidden` (actually, having `.hidden` seems to be Gnome2 devs' idea) but not wildcards in it.

**Reply by tsujan** [reply]
Meanwhile, I found that, after editing `.hidden`, the window/tab should be closed and reopened; a simple reload doesn't work smoothly (sometimes it does, sometimes not). That's because of how `.hidden` is implemented. Will see if it's possible to update it by reloading too.

**Reply by stefonarch** [reply]
Just tested thunar, it supports `.hidden` file but no wildcards.
We could just  go ahead and hoping other filemanager will implement it too ;)

What about dolphin?

**Reply by tsujan** [reply]
> We could just go ahead and hoping other filemanager will implement it too

IMHO, we'd better contribute to making the common implementation a standard by not deviating from it.

> What about dolphin?

Dolphin should have adopted it recently. It doesn't support wildcards either.

**Reply by tsujan** [reply]
Oh, now I remember: all GLib-based file managers automatically support it (I mean `.hidden`) because GLib does.

And GLib doesn't support wildcards in `.hidden`.

**Reply by tsujan** [reply]
As for unreliable reloading when `.hidden` is changed (https://github.com/lxqt/lxqt/discussions/2025#discussioncomment-999451), it's because GLib uses a cache. The user has to wait a few seconds after editing `.hidden` and before reloading the folder. I don't think our code can do anything about ti but I'll try to find a workaround.

**Reply by stefonarch** [reply]
> And GLib doesn't support wildcards in `.hidden`.

So an eventual feature request should go to gnome devs I understand.

Another approach could be include the actual filter in "remember settings for this folder". 
As in my example I'd like to display only "*jpg" and that's what I'm doing now.

**Reply by tsujan** [reply]
> So an eventual feature request should go to gnome devs I understand.

And Gnome devs are known for ignoring feature requests. If they were responsive, I'd file a report about the side-effects of using cache with `.hidden`. I don't make that mistake when I see that a serious report like https://gitlab.gnome.org/GNOME/glib/issues/746 has been ignored for 8 years!

> Another approach could be include the actual filter in "remember settings for this folder".

I'm afraid it's too much, IMO. We should be very cautious about adding more options to per-folder settings, especially after the recursive option. Otherwise, pcmanfm-qt won't be lightweight anymore.

Moreover, `.hidden` is handled by `GLib` at a basic level. It's treated as a file attribute (like name, size, etc.).

**Reply by stefonarch** [reply]
> I'm afraid it's too much, IMO. We should be very cautious about adding more options to per-folder settings, especially after the recursive option.

I agree, thought so too.

**Reply by tsujan** [reply]
Your good idea of a wildcard will remain in my mind. Perhaps, someday, it could be implemented in the core of `libfm-qt` alongside a better cache for `.hidden`.
